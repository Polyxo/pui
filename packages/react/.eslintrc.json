{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:storybook/recommended"
  ],
  "rules": {
    "no-unused-vars": "off",
    "no-console": "error",
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
