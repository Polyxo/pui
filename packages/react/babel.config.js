module.exports = {
  presets: [
    "@babel/preset-env",
    "@babel/preset-react",
    "@babel/preset-typescript",
  ],
  plugins: [
    "dev-expression",
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-syntax-import-meta",
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-export-namespace-from",
    "@babel/plugin-proposal-export-default-from",
  ],
};
