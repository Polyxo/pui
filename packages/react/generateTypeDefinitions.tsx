import { rm, stat, readdir, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { parse } from 'react-docgen-typescript';

const distUrl = path.resolve(__dirname, '../../website/types/');

const options = {
  savePropValueAsString: true,
  propFilter: (prop, component) => {
    if (prop.declarations !== undefined && prop.declarations.length > 0) {
      const hasPropAdditionalDescription = prop.declarations.find(
        (declaration) => {
          return !declaration.fileName.includes('node_modules');
        }
      );

      return Boolean(hasPropAdditionalDescription);
    }

    return true;
  },
};

export default async function walk(directory: string) {
  let fileList: string[] = [];

  const files = await readdir(directory);
  for (const file of files) {
    const p = path.join(directory, file);
    if ((await stat(p)).isDirectory()) {
      fileList = [...fileList, ...(await walk(p))];
    } else {
      fileList.push(p);
    }
  }

  const tsxFileList = fileList.filter(
    (file) => file.endsWith('.tsx') && !file.includes('stories')
  );

  await Promise.all(
    tsxFileList.map(async (file) => {
      const propTypes = JSON.parse(JSON.stringify(parse(file, options)));

      try {
        await mkdir(path.dirname(path.join(distUrl, file)), {
          recursive: true,
        });
      } catch (error) {
        console.log('collectTypes error', error);
      }

      console.log(
        'propTypes',
        path.join(distUrl, file.replace('.tsx', '.json'))
      );
      await writeFile(
        path.join(distUrl, file.replace('.tsx', '.json')),
        JSON.stringify(propTypes, null, 2)
      );
      return null;
    })
  );

  return fileList;
}

async function startGenerateTypes() {
  await rm(distUrl, { recursive: true, force: true });

  await walk('./src');
  process.exit();
}

startGenerateTypes();
