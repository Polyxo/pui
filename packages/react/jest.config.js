module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  transform: {
    "^.+\\.tsx?$": "ts-jest",
  },

  roots: ["<rootDir>/src"],
  testMatch: [
    //"**/__tests__/**/*.+(ts|tsx|js)",
    "**/?(*.)+(spec|test).+(ts|tsx|js)",
  ],
};
