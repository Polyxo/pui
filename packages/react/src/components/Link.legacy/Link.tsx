import React, { PropsWithChildren } from 'react';
import classNames from 'classnames';
import useSettings from '../../hooks/useSettings';

/** Links are used as navigational elements. They may appear on their own, within a sentence or paragraph, or directly following the content. */
interface LinkProps {
  href?: string;
  inline?: boolean;
  className?: string;
}

const Link: React.FC<PropsWithChildren<LinkProps>> = ({
  children,
  className,
  href,
  inline,
  ...other
}) => {
  const { prefix } = useSettings();

  const classes = classNames(
    {
      [`${prefix}--link`]: true,
      [`${prefix}--link--inline`]: inline,
    },
    className
  );
  return (
    <a href={href} className={classes} {...other}>
      {children}
    </a>
  );
};

export default Link;
