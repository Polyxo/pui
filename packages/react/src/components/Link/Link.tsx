import { PropsWithChildren } from 'react';
import * as React from 'react';
import classnames from 'classnames';
import useSettings from '../../hooks/useSettings';
import type { LinkHTMLAttributes } from 'react';

export interface LinkProps extends React.LinkHTMLAttributes {
  /**
   * Adds an underline to the link element to better indicate that it is clickable in continuous text. @design
   */
  inline?: boolean;
  /**
   * Disables the link element. @design
   */
  disabled?: boolean;
  /**
   * Indicates that the link has been visited.
   */
  visited?: boolean;
  /**
   * Choose a size for the link element @design
   */
  size?: 'sm' | 'md' | 'lg';
  /**
   * Use an icon with the link element @design
   */
  icon?: React.node;
  //write other extra props here...
}

const Link: React.FC<LinkProps> = React.forwardRef(
  (
    {
      children,
      className,
      href,
      disabled,
      inline,
      visited,
      icon: Icon,
      size,
      ...other
    },
    ref
  ) => {
    const { prefix } = useSettings();

    const classNames = classnames(`${prefix}--link`, className, {
      [`${prefix}--link--disabled`]: disabled,
      [`${prefix}--link--inline`]: inline,
      [`${prefix}--link--visited`]: visited,
      [`${prefix}--link--${size}`]: size,
    });

    const Tag = disabled ? 'p' : 'a';
    const rel = other.target === '_blank' ? 'noopener' : null;
    return (
      <Tag
        href={disabled ? null : href}
        className={classNames}
        rel={rel}
        ref={ref}
        {...other}>
        {children}
        {!inline && Icon && (
          <div className={`${prefix}--link__icon`}>
            <Icon />
          </div>
        )}
      </Tag>
    );
  }
);

Link.displayName = 'Link';

export default Link;
