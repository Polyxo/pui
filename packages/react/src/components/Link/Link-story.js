/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable no-console */

import React from 'react';
// import { action } from '@storybook/addon-actions';
//import { withKnobs, text, boolean, select } from '@storybook/addon-knobs';
import { Download16 } from '@un/icons-react';
import Link from '../Link';
// import mdx from './Link.mdx';

// const sizes = {
//   'Small  (sm)': 'sm',
//   'Medium (md) - default': undefined,
//   'Large  (lg)': 'lg',
// };

export default {
  title: 'Components/Link',
  parameters: {
    component: Link,
    docs: {
      //  page: mdx,
    },
  },
};

export const _Default = (args) => (
  <Link href="http://www.carbondesignsystem.com" {...args}>
    Link
  </Link>
);

_Default.story = {
  name: 'Link',
};

// export const PairedWithIcon = () => (
//   <Link href="http://www.carbondesignsystem.com" renderIcon={Download16}>
//     Download
//   </Link>
// );
