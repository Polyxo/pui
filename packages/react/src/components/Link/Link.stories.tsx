import React from "react";
import Link from ".";

export default {
  title: "Components/UI Elements/Link",
  component: Link,
  parameters: {
    componentSubtitle: "Component",
    status: "released",
  },
};

export const LinkDefault = (args) => <Link {...args} />;

LinkDefault.args = { children: "WFP Homepage", href: "https://www.wfp.org" };

export const States = (args) => (
  <>
    <Link {...args} className="wfp--link">
      :default
    </Link>
    <Link {...args} className="wfp--link--hover">
      :hover
    </Link>
    <Link {...args} className="wfp--link--focus">
      :focus
    </Link>
    <Link {...args} className="wfp--link--active">
      :active
    </Link>
    <Link {...args} className="wfp--link--visited">
      :visited
    </Link>
    <Link {...args} disabled>
      :disabled
    </Link>
  </>
);

States.args = { ...LinkDefault.args, style: { marginRight: "1rem" } };

States.story = {
  parameters: {
    docs: {
      storyDescription: `The link component only uses minimalistic styles to indicate different states. By default, the link component does not use a visited style. Visited links indicate that a user has already opened the link so they can be a helpful indicator during task completion. Visited styles should be used sparingly because they often clutter the the page and add further visual noise as users are trying to navigate a product. They can be used if it is important that a user knows they have already clicked on a link.
      `,
    },
  },
};

export const LinkSolid = (args) => (
  <div
    style={{
      width: "100%",
      height: "70px",
      backgroundColor: "#04193b",
      padding: "20px",
    }}
  >
    <Link {...args} className="wfp--link--solid">
      :default
    </Link>
    <Link {...args} className="wfp--link--solid--hover">
      :hover
    </Link>

    <Link {...args} className="wfp--link--solid--active">
      :active
    </Link>
    <Link {...args} className="wfp--link--solid--focus">
      :focus
    </Link>
    <Link {...args} className="wfp--link--solid--visited">
      :visited
    </Link>
    <Link {...args} className="wfp--link--solid" disabled>
      :disabled
    </Link>
  </div>
);

LinkSolid.args = {
  ...LinkDefault.args,
  linkSolid: true,
  style: { marginRight: "1rem" },
};

LinkSolid.story = {
  parameters: {
    docs: {
      storyDescription: `The link component only uses minimalistic styles to indicate different states. By default, the link component does not use a visited style. Visited links indicate that a user has already opened the link so they can be a helpful indicator during task completion. Visited styles should be used sparingly because they often clutter the the page and add further visual noise as users are trying to navigate a product. They can be used if it is important that a user knows they have already clicked on a link.
      `,
    },
  },
};
