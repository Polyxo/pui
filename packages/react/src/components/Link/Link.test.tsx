//import React from "react";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
// import userEvent from "@testing-library/user-event";
import Link from "./Link";

describe("Link Component", () => {
  // Test basic rendering
  it("renders a link", () => {
    render(<Link href="https://example.com">Visit Example</Link>);
    expect(screen.getByRole("link")).toHaveTextContent("Visit Example");
    expect(screen.getByRole("link")).toHaveAttribute(
      "href",
      "https://example.com"
    );
  });

  // Test disabled link
  it("renders a disabled link", () => {
    render(<Link disabled>Disabled Link</Link>);
    expect(screen.getByText("Disabled Link")).toHaveAttribute(
      "aria-disabled",
      "true"
    );
  });

  // Test inline link
  it("renders an inline link", () => {
    render(<Link inline>Inline Link</Link>);
    expect(screen.getByText("Inline Link")).toHaveClass("wfp--link--inline");
  });

  // Test visited link
  it("renders a visited link", () => {
    render(<Link visited>Visited Link</Link>);
    expect(screen.getByText("Visited Link")).toHaveClass("wfp--link--visited");
  });

  // Test different sizes of links
  it("renders a small link", () => {
    render(<Link size="sm">Small Link</Link>);
    expect(screen.getByText("Small Link")).toHaveClass("wfp--link--sm");
  });

  it("renders a medium link", () => {
    render(<Link size="md">Medium Link</Link>);
    expect(screen.getByText("Medium Link")).toHaveClass("wfp--link--md");
  });

  it("renders a large link", () => {
    render(<Link size="lg">Large Link</Link>);
    expect(screen.getByText("Large Link")).toHaveClass("wfp--link--lg");
  });

  // Test link with icon
  it("renders a link with an icon", () => {
    const MockIcon = () => <svg data-testid="mock-icon"></svg>;
    render(<Link icon={<MockIcon />}>Link with Icon</Link>);
    expect(screen.getByTestId("mock-icon")).toBeInTheDocument();
    expect(screen.getByText("Link with Icon")).toBeInTheDocument();
  });

  // Test custom class name
  it("applies custom class names", () => {
    const customClass = "my-custom-class";
    render(<Link className={customClass}>Custom Class Link</Link>);
    expect(screen.getByText("Custom Class Link")).toHaveClass(customClass);
  });

  // Test keyboard interaction
  /*
  TODO: Fix this test https://github.com/testing-library/user-event/issues/820
  it("handles keyboard events", () => {
    render(<Link href="https://example.com">Keyboard Link</Link>);
    userEvent.tab();
    expect(screen.getByText("Keyboard Link")).toHaveFocus();
  });
  */

  // Test rel attribute for security when target is "_blank"
  it("adds rel='noopener' when target='_blank'", () => {
    render(
      <Link href="https://example.com" target="_blank">
        External Link
      </Link>
    );
    expect(screen.getByText("External Link")).toHaveAttribute(
      "rel",
      "noopener"
    );
  });

  // Snapshot test
  it("matches snapshot", () => {
    const { asFragment } = render(
      <Link href="https://example.com">Snapshot Link</Link>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
