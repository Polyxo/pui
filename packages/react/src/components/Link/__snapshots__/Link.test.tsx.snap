// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Component matches snapshot 1`] = `
<DocumentFragment>
  <a
    class="wfp--link"
    href="https://example.com"
  >
    Snapshot Link
  </a>
</DocumentFragment>
`;
